src/client/automation/playback/type/type-char.js

```diff
@@ -176,10 +177,37 @@ function _typeCharToTextEditable (element, text) {
     eventSimulator.input(element);
 }
 
-export default function (element, text) {
+function _typeTextToNonTextEditable (element, text, caretPos) {
+    if (caretPos !== null) {
+        var inputValue = element.value;
+
+        for (var i = 0; i < text.length; i++) {
+            if (inputValue[caretPos])
+                inputValue = replaceCharAt(inputValue, caretPos, text[i]);
+            else
+                inputValue += text[i];
+
+            caretPos++;
+        }
+
+        element.value = inputValue;
+    }
+    else
+        element.value = text;
+
+    eventSimulator.change(element);
+    eventSimulator.input(element);
+}
+
+export default function (element, text, caretPos) {
     if (domUtils.isContentEditableElement(element))
         _typeCharToContentEditable(element, text === ' ' ? String.fromCharCode(160) : text);
 
-    if (domUtils.isTextEditableElementAndEditingAllowed(element))
-        _typeCharToTextEditable(element, text);
+    if (!domUtils.isElementReadOnly(element)) {
+        if (domUtils.isTextEditableElement(element))
+            _typeCharToTextEditable(element, text);
+
+        else
+            _typeTextToNonTextEditable(element, text, caretPos);
```

Comment:
potentially we can have an error here, if element is not `input` and it don't have `value`